go/vt/vttablet/tabletserver/schema/historian.go

```diff
@@ -0,0 +1,286 @@
+	GetTableForPos(tableName sqlparser.TableIdent, pos string) *binlogdatapb.MinimalTable
+	Open() error
+	Close()
+	SetTrackSchemaVersions(val bool)
+}
+
+// HistoryEngine exposes only those schema.Engine APIs that are needed by the HistorianSvc
+// We define an interface to facilitate a mock for testing
+var _ HistoryEngine = (*Engine)(nil)
+
+// HistoryEngine defines the interface which is implemented by schema.Engine for the historian to get current schema
+// information from the database
+type HistoryEngine interface {
+	Open() error
+	Close()
+	Reload(ctx context.Context) error
+	GetConnection(ctx context.Context) (*connpool.DBConn, error)
+	GetSchema() map[string]*Table
+}
+
+// TrackedSchema has the snapshot of the table at a given pos (reached by ddl)
+type TrackedSchema struct {
+	schema map[string]*binlogdatapb.MinimalTable
+	pos    mysql.Position
+	ddl    string
+}
+
+var _ Historian = (*HistorianSvc)(nil)
+
+// HistorianSvc implements the Historian interface by calling schema.Engine for the underlying schema
+// and supplying a schema for a specific version by loading the cached values from the schema_version table
+// The schema version table is populated by the Tracker
+type HistorianSvc struct {
+	he                  HistoryEngine
+	lastID              int64
+	schemas             []*TrackedSchema
+	mu                  sync.Mutex
+	trackSchemaVersions bool
+	lastSchema          map[string]*binlogdatapb.MinimalTable
+	isOpen              bool
+}
+
+// Reload gets the latest schema from the database
+func (h *HistorianSvc) Reload(ctx context.Context) error {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	return h.reload(ctx)
+}
+
+// NewHistorian creates a new historian. It expects a schema.Engine instance
+func NewHistorian(he HistoryEngine) *HistorianSvc {
+	sh := HistorianSvc{he: he, lastID: 0, trackSchemaVersions: true}
+	return &sh
+}
+
+// SetTrackSchemaVersions can be used to turn on/off the use of the schema_version history in the historian
+// Only used for testing
+func (h *HistorianSvc) SetTrackSchemaVersions(val bool) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.trackSchemaVersions = val
+}
+
+// readRow converts a row from the schema_version table to a TrackedSchema
+func (h *HistorianSvc) readRow(row []sqltypes.Value) (*TrackedSchema, int64, error) {
+	id, _ := evalengine.ToInt64(row[0])
+	pos, err := mysql.DecodePosition(string(row[1].ToBytes()))
+	if err != nil {
+		return nil, 0, err
+	}
+	ddl := string(row[2].ToBytes())
+	timeUpdated, _ := evalengine.ToInt64(row[3])
+	sch := &binlogdatapb.MinimalSchema{}
+	if err := proto.Unmarshal(row[4].ToBytes(), sch); err != nil {
+		return nil, 0, err
+	}
+	log.Infof("Read tracked schema from db: id %d, pos %v, ddl %s, schema len %d, time_updated %d \n",
+		id, mysql.EncodePosition(pos), ddl, len(sch.Tables), timeUpdated)
+
+	tables := map[string]*binlogdatapb.MinimalTable{}
+	for _, t := range sch.Tables {
+		tables[t.Name] = t
+	}
+	trackedSchema := &TrackedSchema{
+		schema: tables,
```

Comment:
Since `sch` doesn't escape out of this function, it's better to directly assign `sch.Tables`. No need to copy.

Reply:
sch.Tables is a list, converting to a map.